common/src/main/java/com/frostwire/search/knaben/KnabenSearchPerformer.java

```diff
@@ -0,0 +1,393 @@
+            String detailsUrl = torrentData.getDetailsUrl();
+            if (StringUtils.isNullOrEmpty(detailsUrl)) {
+                detailsUrl = "https://" + getDomainName() + "/torrent/" + infoHash;
+            }
+
+            return new KnabenSearchResult(
+                infoHash,
+                torrentData.getName() + ".torrent",
+                torrentData.getName(),
+                magnetUrl,
+                detailsUrl,
+                torrentData.getSize(),
+                torrentData.getCreationTime(),
+                torrentData.getSeeds()
+            );
+        } catch (Exception e) {
+            LOG.warn("Failed to parse torrent JSON object: " + e.getMessage());
+            return null;
+        }
+    }
+
+    private JsonArray findJsonArrayInResponse(JsonElement root) {
+        if (root.isJsonArray()) {
+            return root.getAsJsonArray();
+        }
+        
+        if (root.isJsonObject()) {
+            // Try common field names for torrent results, including "hits" for Knaben API
+            String[] arrayFieldNames = {"hits", "torrents", "results", "data", "items", "response"};
+            for (String fieldName : arrayFieldNames) {
+                if (root.getAsJsonObject().has(fieldName) && root.getAsJsonObject().get(fieldName).isJsonArray()) {
+                    return root.getAsJsonObject().getAsJsonArray(fieldName);
+                }
+            }
+        }
+        
+        return null;
+    }
+
+    private String validateAndNormalizeInfoHash(String infoHash) {
+        if (StringUtils.isNullOrEmpty(infoHash)) {
+            return null;
+        }
+        
+        // Remove any whitespace and convert to lowercase
+        infoHash = infoHash.trim().toLowerCase();
+        
+        // Valid hex info hash should be exactly 40 characters and contain only hex characters
+        if (infoHash.length() == 40 && infoHash.matches("^[a-f0-9]{40}$")) {
+            return infoHash;
+        }
+        
+        // Check if it's base32 encoded (32 characters, uppercase)
+        String upperInfoHash = infoHash.toUpperCase();
+        if (upperInfoHash.length() == 32 && upperInfoHash.matches("^[A-Z2-7]{32}$")) {
+            // BitTorrent magnet links often use base32 encoding for info hashes
+            return upperInfoHash;
+        }
+        
+        // Try to handle other formats that might need cleanup
+        // Remove any non-hex characters and try again
+        String cleanedHash = infoHash.replaceAll("[^a-f0-9]", "");
+        if (cleanedHash.length() == 40 && cleanedHash.matches("^[a-f0-9]{40}$")) {
+            return cleanedHash;
+        }
+        
+        return null;
+    }
+
+    @Override
+    public boolean isCrawler() {
+        return false;
+    }
+
+    // Inner static classes to map JSON response structures
+    
+    /**
+     * Flexible JSON structure for torrent data that handles multiple field naming conventions
+     */
+    public static class TorrentData {
```

Comment:
I might be very late to this PR, but I happened to come across this while looking through FrostWire's search performers in attempting to do some refactoring. I simply have on question:

What in the absolute _fuck_ is this class? (And by that I mean `TorrentData`).

Although I might not have seen too many of FrostWire's search performers, I get the general impression that these classes shouldn't be _this_ big. Based on the general pattern of this class, I feel like AI did something wrong here.

I'm going to have to intervene on this one...

Reply:
<img width="482" height="747" alt="Screenshot 2025-10-03 at 8 47 22 PM" src="https://github.com/user-attachments/assets/f6fc28dc-4d34-46c9-ace9-879341007c41" />

Reply:
my guess is that Knaben Search must be a mix of multiple torrent databases internally

Reply:
Hmm :thinking:

I'll see if I can find any way to clean this up. If there is, you should see a PR for that soon! :smile: